psi4/src/psi4/libmints/oeprop.cc

```diff
@@ -1784,6 +1788,445 @@ PopulationAnalysisCalc::compute_lowdin_charges(bool print_output) {
+        for(int atom = 0; atom < num_atoms; atom++) {
+            if (delta_rho_atoms_0[atom] > delta_rho_max_0) delta_rho_max_0 = delta_rho_atoms_0[atom]; 
+        }
+
+        // Update populations and widths
+        for (int atom = 0; atom < num_atoms; atom++) {
+            for (int m = 0; m < mA[atom]; m++) {
+                Nai[atom][m] = Nai_next[atom][m];
+                Sai[atom][m] = Sai_next[atom][m];
+            }
+        }
+
+        outfile->Printf("   @MBIS iter %3d:  %.3e\n", iter, delta_rho_max_0);
+
+        if (delta_rho_max_0 < CONV) {
+            if (print_output) outfile->Printf("  MBIS Atomic Density Converged\n\n");
+            is_converged = true;
+            break;
+        }
+
+        iter += 1;
+    }
+
+    if (!is_converged) throw ConvergenceError<int>("MBIS", MAX_ITER, CONV, delta_rho_max_0, __FILE__, __LINE__);
+
+    // => Post-Processing <= //
+
+    // Final density calculation afiter convergence
+    for (int point = 0; point < total_points; point++) {
+        rho_0_points[point] = 0.0;
+        for (int atom = 0; atom < num_atoms; atom++) {
+            rho_a_0_points[atom][point] = 0.0;
+            for (int m = 0; m < mA[atom]; m++) {
+                rho_a_0_points[atom][point] += PopulationAnalysisCalc::rho_ai_0(Nai[atom][m], Sai[atom][m], distances[atom][point]);
+            }
+            rho_0_points[point] += rho_a_0_points[atom][point];
+        }
+    }
+      
+    // Atomic density, as defined in Equation 5 of Verstraelen et al.
+    std::vector<std::vector<double>> rho_a(num_atoms, std::vector<double>(total_points, 0.0));
+    for (int atom = 0; atom < num_atoms; atom++) {
+        for (int point = 0; point < total_points; point++) {
+            rho_a[atom][point] = -rho[point] * rho_a_0_points[atom][point] / rho_0_points[point];
+        }
+    }
+
+    // Kronecker Delta
+    std::vector<std::vector<double>> k_delta = {{1.0, 0.0, 0.0}, {0.0, 1.0, 0.0}, {0.0, 0.0, 1.0}};
+
+    // Cartesian atomic multipoles (Horton convention)
+    std::vector<double> monopoles(num_atoms, 0.0);
+    std::vector<std::vector<double>> dipoles(num_atoms, std::vector<double>(3, 0.0));
+    std::vector<std::vector<std::vector<double>>> quadrupoles(num_atoms, std::vector<std::vector<double>>(3, std::vector<double>(3, 0.0)));
+    std::vector<std::vector<std::vector<std::vector<double>>>> octupoles(num_atoms, std::vector<std::vector<std::vector<double>>>(3, std::vector<std::vector<double>>(3, std::vector<double>(3, 0.0))));
+
+    // Cartesian atomic multipoles (convention from Anthony Stone's "The Theory of Intermolecular Forces, Second Edition")
+    std::vector<std::vector<std::vector<double>>> stone_quadrupoles(num_atoms, std::vector<std::vector<double>>(3, std::vector<double>(3, 0.0)));
+    std::vector<std::vector<std::vector<std::vector<double>>>> stone_octupoles(num_atoms, std::vector<std::vector<std::vector<double>>>(3, std::vector<std::vector<double>>(3, std::vector<double>(3, 0.0))));
```

Comment:
we should probably chat about flat/shaped, unique/redundant arrays. Psi and qcsk are committed to returning full, sensibly-shaped arrays to the user upon `variable("* OCTUPOLE")`, etc. that can be mindlessly indexed as `[0][1][1]`. But I'm sure you've met the restriction of `set_array_variable` only taking 2d arrays. The scheme I started for the molecular multipoles is to compute flat+unique here in oeprop, then let https://github.com/psi4/psi4/blob/master/psi4/driver/p4util/python_helpers.py#L638-L670 and other plumper/compressor machinery trigger human-friendly forms. This case is of course different b/c per-atom. 

So we should talk to make sure user gets redundant arrays and to see if both forms needed c-side.